src/ppx/application.cpp

```diff
@@ -1391,6 +1413,94 @@ float2 Application::GetNormalizedDeviceCoordinates(int32_t x, int32_t y) const
     return ndc;
 }
 
+metrics::Run* Application::StartMetricsRun(const char* pName)
+{
+    PPX_ASSERT_MSG(mSettings.useMetrics, "Application::Settings::useMetrics must be set to true before using the metrics capabilities");
+    if (!mSettings.useMetrics) {
+        return nullptr;
+    }
+
+    PPX_ASSERT_MSG(mMetrics.pCurrentRun == nullptr, "a run is already active; stop it before starting another one");
+    mMetrics.pCurrentRun = mMetrics.manager.AddRun(pName);
+
+    // Add default metrics to every single run
+    {
+        metrics::MetricMetadata metadata = {};
+        metadata.name                    = "cpu_frame_time";
+        metadata.unit                    = "ms";
+        metadata.interpretation          = metrics::MetricInterpretation::LOWER_IS_BETTER;
+        mMetrics.pCpuFrameTime           = mMetrics.pCurrentRun->AddMetric<metrics::MetricGauge>(metadata);
+    }
+    {
+        metrics::MetricMetadata metadata = {};
+        metadata.name                    = "framerate";
+        metadata.unit                    = "";
+        metadata.interpretation          = metrics::MetricInterpretation::HIGHER_IS_BETTER;
+        mMetrics.pFramerate              = mMetrics.pCurrentRun->AddMetric<metrics::MetricGauge>(metadata);
+    }
+    {
+        metrics::MetricMetadata metadata = {};
+        metadata.name                    = "frame_count";
+        metadata.unit                    = "";
+        metadata.interpretation          = metrics::MetricInterpretation::NONE;
+        mMetrics.pFrameCount             = mMetrics.pCurrentRun->AddMetric<metrics::MetricCounter>(metadata);
+    }
+
+    mMetrics.resetFramerateTracking = true;
+    return mMetrics.pCurrentRun;
+}
+
+void Application::StopMetricsRun()
```

Comment:
I use it locally to test the metrics.
It just stops the current run, so it should not save the report.
You could have several runs in a single BigWheels execution.
Using the benchmark as an example:
1. Configure some knobs
2. Start a run
3. Record for 10 secs
4. Stop the run
5. Repeat 1 to 4 several times with different run names
6. Exit the app (the report with all runs is exported)
Let me know if that's unclear.

Reply:
Originally, the metrics were developed with automated tests as a use case. In this context, having automatic runs and a dump of the report at the end makes sense. It is also good to test if the thing actually works locally. But, that's definitely not a limitation.

My expectations is that people using the metrics will change the API to match their needs. I think @Slumpwuffle already has this kind of plans.

Basically, the way I see it is that we create a centralized point in application.h to avoid the situation where every single sample reimplements the same mechanics for metrics. But, instead of trying to anticipate every potential use cases, we let people tweak and add what they need along the way.

Does that seem reasonnable?

Reply:
Not sure I follow the whole thing.

> Basically, the way I see it is that we create a centralized point in application.h to avoid the situation where every single sample reimplements the same mechanics for metrics

Sure

> But, instead of trying to anticipate every potential use cases, we let people tweak and add what they need along the way.

This would still be supported since you have the pCurrentRun accessible no?

Maybe I'm missing something, the main point I had was around the logic to prevent concurrent runs. Seems like this could be done by the manager no?

Reply:
OK, misunderstanding then. :)
I thought you were asking for a way to export anytime.

I think I see your point now. There would be only one function that both starts and stops the current run?
I must admit I prefer intentional APIs.

An application using a button could easily manage the state, and they'd need to handle the state anyway to show a proper button label. However, once more applications use the metrics, we can tweak the API to match what's actually convenient or done the most often.

Reply:
I actually like the idea of moving some of this metrics code in Application to Manager, since that's in theory what it's supposed to do, and metrics might be used only by a few applications, so we can keep Application cleaner, and if someone was interested in seeing how metrics work they can find that in a more centralized place

Reply:
No, that's a misunderstanding of the intent of the manager. Applications can already use the manager directly. That's exactly what FishTornado does. In that setting, whether an application creates tons of runs upfront with a bunch of metrics, start recording values, pause, record again, that's all up to the application. A recording flag would have no meaning here, because the application is in full control.

The integration in the Application class is meant to cover the basic use case of most applications: a single active run at a time with a bunch of predefined metrics (which are recorded automatically) + the ability to add custom metrics (which the application records) + the automatic dump to disk. I believe it will actually cover 99% of the cases, but I of course could be wrong.

If that's not clear, maybe we should discuss it live. :)